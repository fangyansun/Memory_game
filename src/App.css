.memory {
  display: flex;
  flex-wrap: wrap;
  width: 360px;
  margin: auto;
  user-select: none;
}
