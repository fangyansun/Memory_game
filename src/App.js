import React, { Component } from 'react'
import shuffle from 'lodash.shuffle'
import Card from './Card'
import GuessCount from './GuessCount'
import './App.css'
import HallOfFame from './HallOfFame'
import HighScoreInput from './HighScoreInput'


const SIDE = 6
const SYMBOLS = '😀🎉💖🎩🐶🐱🦄🐬🌍🌛🌞💫🍎🍌🍓🍐🍟🍿'
const VISUAL_DELAY = 500

class App extends Component {
  state = {
    cards: this.generateCards(),
    currentPair: [],
    guesses: 0,
    hallOfFame: null,
    matchedCardIndices: [],
  }

  generateCards(){
    const result = []
    const size = SIDE*SIDE
    const candidates = shuffle(SYMBOLS)
    while (result.length < size){
      const card = candidates.pop()
      result.push(card,card)
    }
    return shuffle(result)
  }

  getFeedbackForCard(index) {
    const { currentPair, matchedCardIndices } = this.state
    const indexMatched = matchedCardIndices.includes(index)

    if (currentPair.length < 2) {
      return indexMatched || index === currentPair[0] ? 'visible' : 'hidden'
    }

    if (currentPair.includes(index)) {
      return indexMatched ? 'justMatched' : 'justMismatched'
    }

    return indexMatched ? 'visible' : 'hidden'
  }


  // Arrow fx for binding
  handlerCardClick=index => {
    const {cards, currentPair, guesses, matchedCardIndices} = this.state

    if (currentPair.length===2){
      console.log("length = 2")
      return
    }

    if (currentPair.length ===0){
      this.setState({currentPair: [index]})
      return
    }

    if (currentPair.length ===1){
      const newPair = [currentPair[0], index]
      const newGuesses = guesses +1
      const matched = cards[newPair[0]] === cards[newPair[1]]
      console.log(newPair)

      this.setState({currentPair:newPair, guesses: newGuesses})
      if (matched){
        this.setState({matchedCardIndices: [...matchedCardIndices, ...newPair]})
      }
      setTimeout(()=> this.setState({currentPair:[]}), VISUAL_DELAY)
    }
  }

  // Arrow fx for binding
  displayHallOfFame = (hallOfFame) => {
    this.setState({hallOfFame})
  }

  render() {
    const { cards, guesses, hallOfFame, matchedCardIndices} = this.state
    const won = matchedCardIndices.length ===cards.length
    return (
      <div className="memory">
        <GuessCount guesses={guesses} />
        {cards.map((card,index)=>(
          <Card
            card={card}
            feedback={this.getFeedbackForCard(index)}
            index = {index}
            key={index}
            onClick={this.handlerCardClick}
          />

        ))}

        {
          won && (hallOfFame?(
            <HallOfFame entries={hallOfFame}/>
          ):(
            <HighScoreInput guesses={guesses} onStored={this.displayHallOfFame}/>
          ))
        }
      </div>
    )
  }
}

export default App
