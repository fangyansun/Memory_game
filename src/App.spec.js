import { expect } from 'chai'
import React from 'react'
import { shallow } from 'enzyme'
import GuessCount from './GuessCount'

import App from './App'

describe('<App />', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).to.contain(<GuessCount guesses={0} />)
  })

  it('has 36 cards', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('Card')).to.have.length(36)
  })
})
