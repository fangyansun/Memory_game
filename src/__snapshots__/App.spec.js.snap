// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> it should match its reference snapshot 1`] = `""`;
